java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestGenerationUtils.java

```diff
@@ -152,16 +163,280 @@ public static WorkspaceEdit generateTests(List<Object> arguments, IProgressMonit
+            List<String> methodsToTest) throws CoreException {
+        final CompilationUnitChange cuChange = new CompilationUnitChange("", testUnit);
+        final String cuContent = constructNewCU(testUnit, methodsToTest, kind);
+        cuChange.setEdit(new InsertEdit(0, cuContent));
+        return ChangeUtil.convertToWorkspaceEdit(cuChange);
+    }
+
+    /**
+     * Create and get a compilation unit change.
+     */
+    private static WorkspaceEdit createAndGetNewTestCompilationUnit(TestKind kind, ICompilationUnit testUnit,
+            List<String> methodsToTest) throws CoreException {
+        final String cuContent = constructNewCU(testUnit, methodsToTest, kind);
+        final CreateCompilationUnitChange change =
+                new CreateCompilationUnitChange(testUnit, cuContent, "");
+        return ChangeUtil.convertToWorkspaceEdit(change);
+    }
+
+    private static WorkspaceEdit createEditToExistingTestFile(CompilationUnit testRoot, TestKind kind,
+            List<String> methodsToTest, TypeDeclaration typeNode, ITypeBinding typeBinding,
+            IJavaElement insertPosition) throws JavaModelException, CoreException {
+        final String testAnnotation = getTestAnnotation(kind);
+        final List<MethodMetaData> metadata = methodsToTest.stream().map(method -> {
+            final String methodName = getTestMethodName(method);
+            return new MethodMetaData(methodName, testAnnotation);
+        }).collect(Collectors.toList());
+
+        final TextEdit edit = getTextEdit(kind, metadata, testRoot, typeNode, typeBinding, insertPosition);
+        return SourceAssistProcessor.convertToWorkspaceEdit((ICompilationUnit) testRoot.getJavaElement(), edit);
+    }
+
+    private static String constructNewCU(ICompilationUnit testUnit,
+            List<String> methods, TestKind testKind) throws CoreException {
+        final String delimiter = StubUtility.getLineDelimiterUsed(testUnit);
+        final String typeStub = constructTypeStub(testUnit, methods, testKind, delimiter);
+        final String cuContent = constructCUContent(testUnit, testKind, typeStub, delimiter);
+        final String formattedCuStub = CodeFormatterUtil.format(
+                CodeFormatter.K_COMPILATION_UNIT, cuContent, 0, delimiter, testUnit.getJavaProject().getOptions(true));
+        return formattedCuStub;
+    }
+
+    private static String constructCUContent(ICompilationUnit testUnit, TestKind testKind, 
+            String typeContent, String delimiter) throws CoreException {
+        final IPackageFragment packageFragment = (IPackageFragment) testUnit.getParent();
+        final StringBuilder buf = new StringBuilder();
+        if (!packageFragment.isDefaultPackage()) {
+            buf.append("package ")
+                .append(packageFragment.getElementName())
+                .append(";")
+                .append(delimiter)
+                .append(delimiter)
```

Comment:
why 2 delimiters here?

Reply:
I got it, an empty line. BTW `lineDelimeter` would be more informative.

Reply:
or `eol`

Reply:
ok